### DISCUSSION TITLE
Optional Presentation functionality on the navigation-beta branch

### DISCUSSION THREAD

**Discussion by emlynmac** [pointfreeco/swift-composable-architecture#2098] [Beta]
I’m working through updating my app to TCA ReducerProtocol and I’m looking to include the navigation components laid out with Destinations.
I’ve got my project hooked up to TCA navigation-beta branch and have a basic sheet implementation _almost_ working.

I’m using this as a reference for the implementation:
https://github.com/pointfreeco/episode-code-samples/blob/5c17384710ae8ef9bc3457a7d1d9fc7982281ea8/0229-composable-navigation-pt8/Inventory/Inventory/Inventory.swift#L319

The main issue I have is that the `.sheet` view modifier in the episode does not seem to be in the navigation-beta branch:
https://github.com/pointfreeco/episode-code-samples/blob/5c17384710ae8ef9bc3457a7d1d9fc7982281ea8/0229-composable-navigation-pt8/Inventory/Inventory/Navigation.swift#L147

It looks to me like the view modifier should have optional presentation state in this file:

https://github.com/pointfreeco/swift-composable-architecture/blob/5fbce502a0070c32a81d0c7e5a6def86b68bfa73/Sources/ComposableArchitecture/SwiftUI/Sheet.swift#L20

Am I missing something or should this be optional?

If I use the functions from the episode, I'm good to go:
```
extension View {
  func sheet<DestinationState, DestinationAction, ChildState: Identifiable, ChildAction>(
    store: Store<DestinationState?, PresentationAction<DestinationAction>>,
    state toChildState: @escaping (DestinationState) -> ChildState?,
    action fromChildAction: @escaping (ChildAction) -> DestinationAction,
    @ViewBuilder child: @escaping (Store<ChildState, ChildAction>) -> some View
  ) -> some View {
    self.sheet(
      store: store.scope(
        state: { $0.flatMap(toChildState) },
        action: {
          switch $0 {
          case .dismiss:
            return .dismiss
          case let .presented(action):
            return .presented(fromChildAction(action))
          }
        }
      ),
      child: child
    )
  }
  
  func sheet<ChildState: Identifiable, ChildAction>(
      store: Store<ChildState?, PresentationAction<ChildAction>>,
      @ViewBuilder child: @escaping (Store<ChildState, ChildAction>) -> some View
    ) -> some View {
      WithViewStore(store, observe: { $0?.id }) { viewStore in
        self.sheet(
          item: Binding(
            get: { viewStore.state.map { Identified($0, id: \.self) } },
            set: { newState in
              if viewStore.state != nil {
                viewStore.send(.dismiss)
              }
            }
          )
        ) { _ in
          IfLetStore(
            store.scope(
              state: returningLastNonNilValue { $0 },
              action: PresentationAction.presented
            )
          ) { store in
            child(store)
          }
        }
      }
    }
}

func returningLastNonNilValue<A, B>(
  _ f: @escaping (A) -> B?
) -> (A) -> B? {
  var lastValue: B?
  return { a in
    lastValue = f(a) ?? lastValue
    return lastValue
  }
}
```

**Comment by stephencelis** [comment]
@emlynmac Are you caught up on episodes? The sample you link to is from 229, which is _before_ the `PresentationState` property wrapper is introduced. This property wrapper bakes in optionality, and is used in the `navigation-beta` branch.

If you share a full code sample that you're trying to get to work on `navigation-beta` we can try to help steer you in the right direction.

**Reply by emlynmac** [reply]
Thanks @stephencelis. I was not fully caught up...
I realized after looking at my code a little more closely, the magic I was missing. 
Using: `@PresentationState var destination: Destination.State?` in my app state and then the issue in the SwiftUI was to add the missing `$`

```
.sheet(
  store: self.store.scope(
    state: \.$destination,                                 //// This was just \.destination previously
    action: AppRootLogic.Action.destination
  ),
...
```

I've now got the prerelease/1.0 branch in my project and zero deprecation warnings, and everything is happy.
